### PR TITLE
Add missing "Size" field in publications

### PR DESCRIPTION
**Pull request by m-bucher**
Closes #8506
https://pulp.plan.io/issues/8506

### CODE REVIEW

File: pulp_deb/tests/functional/api/test_publish_complex_dists.py [hunk 1]
```diff
@@ -170,4 +173,33 @@ def do_publish(self, expected_values):
         download_content_unit(cfg, distribution.to_dict(), release_file_path)
 
         for package_index_path in expected_values["package_index_paths"]:
-            download_content_unit(cfg, distribution.to_dict(), package_index_path)
+            published = download_content_unit(cfg, distribution.to_dict(), package_index_path)
+            remote = http_get("/".join([remote_body["url"], package_index_path]))
```
Comment (quba42) [inline pulp_deb/tests/functional/api/test_publish_complex_dists.py:177]: I wonder if it would not make more sense to just use the `DEB_COMPLEX_DISTS_FIXTURE_URL` directly here, instead of going through the remote we created?

File: pulp_deb/tests/functional/api/test_publish_complex_dists.py [hunk 2]
```diff
@@ -170,4 +173,33 @@ def do_publish(self, expected_values):
...
+                    pkg[k], orig_pkg[k], "Field '{}' of package '{}' does not match".format(k, name)
+                )
+
+    def parse_package_index(self, pkg_idx):
+        """Parses PackageIndex file-string.
+        Returns a dict of the packages by '<Package>-<Version>'.
+        """
+        packages = {}
+        for package in deb822.Packages.iter_paragraphs(pkg_idx):
+            packages["-".join([package["Package"], package["Version"]])] = package
```
Comment (quba42) [inline pulp_deb/tests/functional/api/test_publish_complex_dists.py:204]: Not sure if Package name and Version alone are enough to guarantee a unique dict key...

File: pulp_deb/tests/functional/api/test_publish_complex_dists.py [hunk 3]
```diff
@@ -170,4 +173,33 @@ def do_publish(self, expected_values):
...
+            published = download_content_unit(cfg, distribution.to_dict(), package_index_path)
+            remote = http_get("/".join([remote_body["url"], package_index_path]))
+            self.assert_equal_package_index(remote, published)
+
+    def assert_equal_package_index(self, orig, new):
+        """In-detail check of two PackageIndex file-strings"""
+        parsed_orig = self.parse_package_index(orig)
+        parsed_new = self.parse_package_index(new)
+
+        self.assertEqual(len(parsed_orig), len(parsed_new))
```
Comment (quba42) [inline pulp_deb/tests/functional/api/test_publish_complex_dists.py:185]: Should we also make sure that `len` is greater 0? Just to make sure the test does not pass trivially because `parse_package_index` returned something empty?

File: pulp_deb/app/tasks/synchronizing.py [hunk 4]
```diff
@@ -622,7 +622,9 @@ async def _handle_package_index(
                 )
                 package_path = os.path.join(self.parsed_url.path, package_relpath)
                 package_da = DeclarativeArtifact(
-                    artifact=Artifact(**_get_checksums(package_paragraph)),
+                    artifact=Artifact(
+                        size=int(package_paragraph["Size"]), **_get_checksums(package_paragraph)
+                    ),
```
Comment (mdellweg) [inline pulp_deb/app/tasks/synchronizing.py:627]: This will add validation of the size of the file when downloading it, i think. Let's hope metadata does not lie.

**Review by mdellweg — COMMENTED** [review]
The code change looks good to me. :+1:
Haven't looked much into the test changes, however.

File: pulp_deb/app/tasks/synchronizing.py (hunk 4 above)
Comment (m-bucher) [reply pulp_deb/app/tasks/synchronizing.py:627]: In a perfect world it would be a good feature to validate the size, because it makes the package-checksum more secure.
However, in our future :crystal_ball: I see an issue that complains about why repository xyz cannot have the correct package size in the package-index file and should be synced nevertheless :roll_eyes:

**Comment by quba42** [comment]
The test fails with: `ModuleNotFoundError: No module named '/usr/local/lib/python3'`
What?!
And this happens in unittests which you did not change... Maybe we need the latest plugin template?!

File: pulp_deb/app/tasks/synchronizing.py (hunk 4 above)
Comment (mdellweg) [reply pulp_deb/app/tasks/synchronizing.py:627]: I am mostly concerned that some repositories will count the final linebreak in metadata files differently. Not sure if this comes from a distant memory, or is just overcautious.

Comment (quba42) [reply pulp_deb/app/tasks/synchronizing.py:627]: I can't think of a better approach than merging and later adding exception handling if it becomes a problem...

File: pulp_deb/tests/functional/api/test_publish_complex_dists.py (hunk 3 above)
Comment (m-bucher) [reply pulp_deb/tests/functional/api/test_publish_complex_dists.py:185]: Given that some of our fixture Package indizes are empty to begin with (e.g. https://fixtures.pulpproject.org/debian-complex-dists/dists/ragnarok-backports/asgard/binary-armeb/Packages).
I do not think we should do this.

Unless we want to store for the package-indizes, how much packages we expect.
But this would make the pulp_deb test-code even more dependant on the fixtures, which IMHO can be a problem, because it ties versions/commits of pulpcore-fixtures close together to versions/commits in pulp_deb.

File: pulp_deb/tests/functional/api/test_publish_complex_dists.py (hunk 1 above)
Comment (m-bucher) [reply pulp_deb/tests/functional/api/test_publish_complex_dists.py:177]: Makes sense :sweat_smile:

File: pulp_deb/tests/functional/api/test_publish_complex_dists.py (hunk 2 above)
Comment (m-bucher) [reply pulp_deb/tests/functional/api/test_publish_complex_dists.py:204]: Agreed, thanks

File: pulp_deb/tests/functional/api/test_publish_complex_dists.py (hunk 3 above)
Comment (quba42) [reply pulp_deb/tests/functional/api/test_publish_complex_dists.py]: I wasn't aware of the empty index... I guess we just leave it as it is.

**Review by quba42 — APPROVED** [review]
I performed one final manual test with `policy="on_demand"` which was now successfull, I am merging.
